ftl/node/builder_test.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import unittest
+import tempfile
+
+from containerregistry.client.v2_2 import docker_image
+
+from ftl.common import context
+from ftl.common import cache
+from ftl.node import builder
+
+class NodeTest(unittest.TestCase):
+    def test_create_package_base_uncached(self):
+        b = builder.Node(None)
+        current_dir = os.path.dirname(__file__)
+        base_config_path = os.path.join(current_dir,
+         "testdata/base_image/config_file")
+        with open(base_config_path, 'r') as reader:
+            base_config = reader.read()
+        ctx = context.Workspace(os.path.join(current_dir,
+         "testdata/node_app"))
+        with builder.Node(ctx) as b:
+            # TODO(aaron-prindle) use fast image format instead of tarball
+            base_image_path = os.path.join(current_dir,
+             "testdata/base_image/distroless-nodejs-latest.tar.gz")
+            with docker_image.FromDisk(base_config, zip([],[]), base_image_path) as base_image:
+                tmp = tempfile.mkdtemp()
+                cash = cache.MockHybridRegistry('fake.gcr.io/google-appengine', tmp)
+                with b.CreatePackageBase(base_image, cash) as _:
+                    # check that the node_app_layer image was added to the cache
+                    self.assertEqual(1, len(cash.GetMap()))
+                    for k in cash.GetMap():
+                        self.assertEqual(str(k).startswith(
+                            "fake.gcr.io/google-appengine/node-package-lock-cache"
+                            ),
+                        True)
+                cash.ResetCacheMiss()
```

Comment:
To make this test easier to read, you might want to move most of the boilerplate into a Setup method, then the testcases themselves can be small. You can probably ditch the restCacheMiss API if you switch to several testcases.

Reply:
still working on this